input-output-hk/mantis
# [ETCM-44] Treasury block reward distribution

# Description

Adapts block reward distribution to the treasury

# Proposed Solution

Follows the specification described in [this](https://input-output.atlassian.net/wiki/spaces/EMU/pages/1681293384/Proto-Treasury#Block-execution) section:
- _If treasury contract wasn't deployed or was selfdestroyed_: follow ETC rules, 100% reward for block to miner
- _If treasury contract was deployed but the miner opted-out_: 80% reward for block to miner
- _If treasury contract was deployed and miner opted-in_: 80% reward for block to miner, 20% reward for treasury


# Testing

## Testing on mainnet

- Mantis should still remain compatible with ETC mainnet and mordor

## Testing on private network

1. Start up a node
  - Gas limit on mainnet genesis isn't enough for the txs being sent below, a value of 0x1388000000 worked for me
  - It requires an account with funds so you should create one and add it to the genesis file
  - Requires configuration of private network: mocked mining on, fast sync off, pruning off, etc
2. Deploy a contract with a function that calls selfdestroy, simpler one that can be used is:
```
pragma solidity ^0.5.0;

contract MockTreasury {
    function shutdown() public {
        // All funds from the contract are burned
        selfdestruct(address(0));
    }
}
```
3. Restart the node after:
 - Configuring the deployed contract address as the treasury address
 - Setting the ECIP1098 block number activation to a more feasible one
4. Mine enough blocks to surpass the ECIP1098 activation block
 - Block reward distribution should behave as spec'ed, both before and after the activation of ECIP1098
5. Call function to selfdestroy the contract, mine further blocks and check that block reward distribution behaves as before ECIP1098 activation

# Pending

- [x] Test on mainnet
- [x] Add useful endpoints examples to insomnia
